pkg/upgrade/upgrade.go

```diff
@@ -0,0 +1,289 @@
+	gitssh "gopkg.in/src-d/go-git.v4/plumbing/transport/ssh"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+type UpgradeValues struct {
+	RepoOrgAndName string
+	Remote         string
+	HeadBranch     string
+	BaseBranch     string
+	CommitMessage  string
+	Name           string
+	Namespace      string
+	ProfileBranch  string
+	ConfigMap      string
+	Out            string
+	ProfileRepoURL string
+	ProfilePath    string
+	GitRepository  string
+	Version        string
+}
+
+func Upgrade(upgradeValues UpgradeValues, w io.Writer) error {
+	config := config.GetConfigOrDie()
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(w, "Checking if entitlement exists...\n")
+
+	entitlement, err := getEntitlement(clientset)
+	if err != nil {
+		return err
+	}
+
+	upgradeValues.Name = "weave-gitops-enterprise"
+	upgradeValues.Namespace = "wego-system"
+	upgradeValues.ProfileBranch = "main"
+	upgradeValues.ConfigMap = ""
+	upgradeValues.Out = "."
+	upgradeValues.ProfileRepoURL = "git@github.com:weaveworks/weave-gitops-enterprise-profiles.git"
+	upgradeValues.ProfilePath = "weave-gitops-enterprise"
+
+	repoURL, err := getRepoURL(upgradeValues.Remote)
+	if err != nil {
+		return err
+	}
+
+	if upgradeValues.RepoOrgAndName == "" {
+		githubRepoPath, err := getRepoOrgAndName(repoURL)
+		if err != nil {
+			return err
+		}
+
+		fmt.Fprintf(w, "Deriving org/repo for PR as %v\n", githubRepoPath)
+		upgradeValues.RepoOrgAndName = githubRepoPath
+	}
+
+	if upgradeValues.GitRepository == "" {
+		gitRepositoryNameNamespace := "wego-system/" + strings.TrimSuffix(filepath.Base(repoURL), ".git")
+		fmt.Fprintf(w, "Deriving name of GitRepository Resource as %v\n", gitRepositoryNameNamespace)
+		upgradeValues.GitRepository = gitRepositoryNameNamespace
+	}
+
+	key := entitlement.Data["deploy-key"]
+
+	localRepo, err := cloneToTempDir("", upgradeValues.ProfileRepoURL, upgradeValues.ProfileBranch, key, w)
+	if err != nil {
+		return err
+	}
+
+	upgradeValues.ProfileRepoURL = localRepo.WorktreeDir()
+
+	installationDirectory, err := addProfile(upgradeValues)
+	if err != nil {
+		return err
+	}
+
+	if err := createPullRequest(upgradeValues, installationDirectory); err != nil {
+		return err
+	}
+
+	fmt.Fprintf(w, "Upgrade pull request created\n")
+
+	return nil
+}
+
+func getRepoOrgAndName(url string) (string, error) {
+	repoEndpoint, err := transport.NewEndpoint(url)
+	if err != nil {
+		return "", err
+	}
+
+	return strings.Trim(strings.TrimSuffix(strings.TrimSpace(repoEndpoint.Path), ".git"), "/"), nil
+}
+
+func cloneToTempDir(parentDir, gitURL, branch string, privKey []byte, w io.Writer) (*GitRepo, error) {
+	fmt.Fprintf(w, "Creating a temp directory...")
+
+	gitDir, err := ioutil.TempDir(parentDir, "git-")
+	if err != nil {
+		return nil, errors.Wrap(err, "TempDir")
```

Comment:
Does it make sense to use `fmt.Errorf` everywhere?

Reply:
Yesss! Thank you for pointing this out 🙂